plugins/mapper-annotated-text/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/AnnotatedTextHighlighter.java

```diff
@@ -0,0 +1,152 @@
+            }
+            if (field.fieldOptions().numberOfFragments() == 0) {
+                // we use a control char to separate values, which is the only char that the custom break iterator
+                // breaks the text on, so we don't lose the distinction between the different values of a field and we
+                // get back a snippet per value
+                CustomSeparatorBreakIterator breakIterator = new CustomSeparatorBreakIterator(MULTIVAL_SEP_CHAR);
+                highlighter = new CustomUnifiedHighlighter(searcher, analyzer, offsetSource, passageFormatter,
+                    field.fieldOptions().boundaryScannerLocale(), breakIterator, fieldValue, field.fieldOptions().noMatchSize());
+                numberOfFragments = fieldValues.size(); // we are highlighting the whole content, one snippet per value
+            } else {
+                //using paragraph separator we make sure that each field value holds a discrete passage for highlighting
+                BreakIterator bi = getBreakIterator(field);
+                highlighter = new CustomUnifiedHighlighter(searcher, analyzer, offsetSource, passageFormatter,
+                    field.fieldOptions().boundaryScannerLocale(), bi,
+                    fieldValue, field.fieldOptions().noMatchSize());
+                numberOfFragments = field.fieldOptions().numberOfFragments();
+            }
+
+            if (field.fieldOptions().requireFieldMatch()) {
+                final String fieldName = highlighterContext.fieldName;
+                highlighter.setFieldMatcher((name) -> fieldName.equals(name));
+            } else {
+                highlighter.setFieldMatcher((name) -> true);
+            }
+
+            Snippet[] fieldSnippets = highlighter.highlightField(highlighterContext.fieldName,
+                highlighterContext.query, hitContext.docId(), numberOfFragments);
+            for (Snippet fieldSnippet : fieldSnippets) {
+                if (Strings.hasText(fieldSnippet.getText())) {
+                    snippets.add(fieldSnippet);
+                }
+            }
+        } catch (IOException e) {
+            throw new FetchPhaseExecutionException(context,
+                "Failed to highlight field [" + highlighterContext.fieldName + "]", e);
+        }
+
+        snippets = filterSnippets(snippets, field.fieldOptions().numberOfFragments());
+
+        if (field.fieldOptions().scoreOrdered()) {
+            //let's sort the snippets by score if needed
+            CollectionUtil.introSort(snippets, (o1, o2) -> Double.compare(o2.getScore(), o1.getScore()));
+        }
+
+        String[] fragments = new String[snippets.size()];
+        for (int i = 0; i < fragments.length; i++) {
+            fragments[i] = snippets.get(i).getText();
+        }
+
+        if (fragments.length > 0) {
+            return new HighlightField(highlighterContext.fieldName, Text.convertFromStringArray(fragments));
+        }
+        return null;
+    }
+    
```

Comment:
A lot of this is just a copy of the parent method; is it worth trying to break it up a bit more and use inheritance to avoid copying so much code?  The only differences are in the `Analyzer` and `PassageFormatter` types, I think?